torch/distributed/device_mesh.py

```diff
@@ -1038,6 +1038,86 @@ def _flatten(
                 dim_names=(mesh_dim_name,) if mesh_dim_name else None,
             )
 
+        def _unflatten(
+            self,
+            dim: Union[int, str],
+            mesh_sizes: tuple[int, ...],
+            mesh_dim_names: tuple[str, ...],
+        ) -> "DeviceMesh":
+            """
+            Returns a DeviceMesh by unflatten the current DeviceMesh.
+
+            This api can be used to unflatten a N-D DeviceMesh into N-1+len(mesh_sizes)-D meshes or submeshes.
+            The dim is the dimension to be unflattened which can be either a string or an integer.
+            
+            The mesh_sizes is a tuple which specifies the shape of the mesh unflatten into for the given dim.
+            The mesh_dim_names is a list of strings which specifies the names of the dimensions of the mesh unflatten into.
+            Its length must match the length of mesh_sizes.
+            
+            For example, if we have a 1D mesh DeviceMesh([0, 1, 2, 3, 4, 5, 6, 7], mesh_dim_names=("world")),
+            calling mesh_1d._unflatten(0, (2, 2, 4), ["dp", "pp", "tp"]) will create a 3D mesh
+            DeviceMesh([[[0, 1], [2, 3]], [[4, 5], [6, 7]]], mesh_dim_names=("dp", "cp", "tp")).
+            
+            After calling the unflatten, to access the unflattened dimension in mesh_1d, one can use the
```

Comment:
is this PR actually doing what this comment says?  (does it allow you to slice the unflattened dim out of the root, or is that just copypaste from the old PR)  I am not seeing the code that would modify the self mesh, so i think its just stale comments.

Reply:
I directly copied from old PR, I will update it later.